src/helpers.c

```diff
@@ -94,12 +91,20 @@ const char *to_voltage(int16_t i) {
     return v;
 }
 
-
-int16_t bit_reverse(int16_t unreversed) {
-    int16_t NO_OF_BITS = 16;
+int16_t bit_reverse(int16_t unreversed, int8_t bits_to_reverse) {
     int16_t reversed = 0;
-    for (int i = 0; i < NO_OF_BITS; i++) {
-        if ((unreversed & (1 << i))) reversed |= 1 << ((NO_OF_BITS - 1) - i);
+    for (int i = 0; i < bits_to_reverse; i++) {
+        if ((unreversed & (1 << i)))
+            reversed |= 1 << ((bits_to_reverse - 1) - i);
     }
     return reversed;
 }
+
+int16_t rev_bitstring_to_int(const char* token) {
+    int8_t length = strlen(token);
+    int16_t value = 0;
+    for (int8_t i = 0; i < length; i++) {
+        if (token[i] == '1') { value += 1 << i; }
```

Comment:
i'm not sure this will work properly for negative values, i would use `value |=` instead of `value +=` to be on the safe side.

Reply:
i did a quick test and it appears it works properly either way. i would still test with various negative values just to be sure, i used visual studio c++ to test, so this might behave differently.

Reply:
Unless I'm mistaken, the | vs + doesn't really matter, since we're just picking up bits set in a string comprised of zeros and ones at this point, thus always generating a positive number here. It works because the tt later chooses to interpret large numbers as negative. I tried manually wrapping value after INT16_MAX (like the tt works) just after this just before return, and it makes no difference, it works as expected either way. In all my tests on the tt it behaves as expected... I don't see any reason why this should be changed, but I'm eager to be proven wrong!

Reply:
but it's not generating positive numbers only. we are adding to a signed 16 bit integer, so that means that once it gets to `1 << 15` that will be interpreted as `-32768` (due to the implicit casting, since it's adding to a signed integer). i think what happens is that due to how two's complement numbers work it ends up giving the correct result either way - with `+=` or `|=` (which might test confirmed, but i tested with c++ instead of c and using a different compiler, so the behaviour might be different). in any case, as long as it's providing the correct result it doesn't matter which way it's done.

Reply:
Agreed!